File: python/src/tree.py
import math
from typing import Optional, Dict, List
from graphviz import Digraph
import os


class Node:
    def __init__(self, value, left: Optional['Node'] = None, right: Optional['Node'] = None):
        self.value = value
        self.left = left
        self.right = right


def build_tree(rpn: List) -> Optional[Node]:
    """Build binary expression tree from RPN"""
    stack = []
    operators = {'+', '-', '*', '/', '^', '√'}
    
    for token in rpn:
        # Check if token is an operator
        if token in operators:
            # Unary operators (√)
            if token == '√':
                if not stack:
                    raise ValueError("Invalid expression: sqrt without operand")
                operand = stack.pop()
                stack.append(Node(token, operand, None))
            # Binary operators
            else:
                if len(stack) < 2:
                    raise ValueError(f"Invalid expression: {token} needs two operands")
                right = stack.pop()
                left = stack.pop()
                stack.append(Node(token, left, right))
        # Operand (number or variable)
        else:
            stack.append(Node(token))
    
    if len(stack) != 1:
        raise ValueError("Invalid RPN: too many operands")
    
    return stack[0]


def print_tree(node: Optional[Node], prefix: str = "", is_tail: bool = True):
    """Print tree structure with proper formatting"""
    if node is None:
        return
    
    print(prefix + ("└── " if is_tail else "├── ") + str(node.value))
    
    children = []
    if node.left:
        children.append(node.left)
    if node.right:
        children.append(node.right)
    
    for i, child in enumerate(children):
        is_last = (i == len(children) - 1)
        extension = "    " if is_tail else "│   "
        print_tree(child, prefix + extension, is_last)


def evaluate_tree(node: Optional[Node], variables: Dict[str, float] = None) -> float:
    """Evaluate expression tree with variable substitution"""
    if variables is None:
        variables = {}
    
    if node is None:
        raise ValueError("None node encountered")
    
    # Number
    if isinstance(node.value, (int, float)):
        return float(node.value)
    
    # Variable (string that's not an operator)
    if isinstance(node.value, str):
        operators = {'+', '-', '*', '/', '^', '√'}
        if node.value not in operators:
            if node.value not in variables:
                raise ValueError(f"Undefined variable: {node.value}")
            return variables[node.value]
    
    # Evaluate children
    left_val = evaluate_tree(node.left, variables) if node.left else None
    right_val = evaluate_tree(node.right, variables) if node.right else None
    
    # Apply operator
    op = node.value
    if op == '+':
        return left_val + right_val
    elif op == '-':
        return left_val - right_val
    elif op == '*':
        return left_val * right_val
    elif op == '/':
        if right_val == 0:
            raise ZeroDivisionError("Division by zero")
        return left_val / right_val
    elif op == '^':
        return math.pow(left_val, right_val)
    elif op == '√':
        if left_val < 0:
            raise ValueError("Square root of negative number")
        return math.sqrt(left_val)
    else:
        raise ValueError(f"Unknown operator: {op}")


def visualize_tree(root: Optional[Node], filename: str = "expression_tree", output_dir: str = "../output"):
    """
    Generate tree visualization using Graphviz and save to output directory.
    
    Args:
        root: Root node of expression tree
        filename: Name of output file (without extension)
        output_dir: Directory to save output (relative to src/)
    """
    if root is None:
        print("Empty tree, no visualization generated")
        return
    
    # Create output directory if it doesn't exist
    script_dir = os.path.dirname(os.path.abspath(__file__))
    full_output_dir = os.path.join(script_dir, output_dir)
    os.makedirs(full_output_dir, exist_ok=True)
    
    # Create Graphviz object
    dot = Digraph(comment='Expression Tree')
    dot.attr(rankdir='TB')
    dot.attr('node', shape='circle', style='filled', fillcolor='lightblue')
    
    operators = {'+', '-', '*', '/', '^', '√'}
    
    def add_nodes(node, parent_id=None):
        if node is None:
            return
        
        node_id = str(id(node))
        label = str(node.value)
        
        # Color coding for different node types
        if isinstance(node.value, (int, float)):
            dot.node(node_id, label, fillcolor='lightgreen')
        elif isinstance(node.value, str) and node.value not in operators:
            dot.node(node_id, label, fillcolor='lightyellow')  # Variables
        else:
            dot.node(node_id, label, fillcolor='lightblue')  # Operators
        
        if parent_id:
            dot.edge(parent_id, node_id)
        
        add_nodes(node.left, node_id)
        add_nodes(node.right, node_id)
    
    add_nodes(root)
    
    # Save to output directory
    output_path = os.path.join(full_output_dir, filename)
    dot.render(output_path, format='png', cleanup=True)
    
    print(f"✓ Tree visualization saved: {output_path}.png")

File: cpp/src/tree.cpp
#include "tree.hpp"
#include <stack>
#include <stdexcept>
#include <cmath>
#include <iostream>
#include <set>

// Operator set for checking
const std::set<std::string> OPERATORS = {"+", "-", "*", "/", "^", "√"};

// Node constructors
Node::Node(double val) : value(val), left(nullptr), right(nullptr) {}

Node::Node(const std::string& val) : value(val), left(nullptr), right(nullptr) {}

Node::Node(const std::string& op, std::unique_ptr<Node> l, std::unique_ptr<Node> r)
    : value(op), left(std::move(l)), right(std::move(r)) {}

// Check if node is operator
bool Node::isOperator() const {
    if (std::holds_alternative<std::string>(value)) {
        return OPERATORS.count(std::get<std::string>(value)) > 0;
    }
    return false;
}

// Check if node is number
bool Node::isNumber() const {
    return std::holds_alternative<double>(value);
}

// Check if node is variable
bool Node::isVariable() const {
    if (std::holds_alternative<std::string>(value)) {
        std::string str = std::get<std::string>(value);
        return OPERATORS.count(str) == 0;
    }
    return false;
}

// Get string representation
std::string Node::toString() const {
    if (std::holds_alternative<double>(value)) {
        return std::to_string(std::get<double>(value));
    } else {
        return std::get<std::string>(value);
    }
}

// Build tree from RPN
std::unique_ptr<Node> buildTree(const std::vector<TokenValue>& rpn) {
    std::stack<std::unique_ptr<Node>> stack;
    
    for (const auto& token : rpn) {
        // Check if it's an operator
        bool isOp = false;
        std::string opStr;
        
        if (std::holds_alternative<std::string>(token)) {
            opStr = std::get<std::string>(token);
            isOp = OPERATORS.count(opStr) > 0;
        }
        
        if (isOp) {
            // Unary operator (√)
            if (opStr == "√") {
                if (stack.empty()) {
                    throw std::runtime_error("Invalid expression: sqrt without operand");
                }
                auto operand = std::move(stack.top());
                stack.pop();
                stack.push(std::make_unique<Node>(opStr, std::move(operand), nullptr));
            }
            // Binary operators
            else {
                if (stack.size() < 2) {
                    throw std::runtime_error("Invalid expression: operator needs two operands");
                }
                auto right = std::move(stack.top());
                stack.pop();
                auto left = std::move(stack.top());
                stack.pop();
                stack.push(std::make_unique<Node>(opStr, std::move(left), std::move(right)));
            }
        }
        // Operand (number or variable)
        else {
            if (std::holds_alternative<double>(token)) {
                stack.push(std::make_unique<Node>(std::get<double>(token)));
            } else {
                stack.push(std::make_unique<Node>(std::get<std::string>(token)));
            }
        }
    }
    
    if (stack.size() != 1) {
        throw std::runtime_error("Invalid RPN: too many operands");
    }
    
    return std::move(stack.top());
}

// Print tree with proper formatting
void printTree(const Node* node, const std::string& prefix, bool isTail) {
    if (node == nullptr) {
        return;
    }
    
    std::cout << prefix << (isTail ? "└── " : "├── ") << node->toString() << "\n";
    
    // Collect children
    std::vector<const Node*> children;
    if (node->left) children.push_back(node->left.get());
    if (node->right) children.push_back(node->right.get());
    
    // Print children
    for (size_t i = 0; i < children.size(); ++i) {
        bool isLast = (i == children.size() - 1);
        std::string extension = isTail ? "    " : "│   ";
        printTree(children[i], prefix + extension, isLast);
    }
}

// Evaluate tree recursively
double evaluateTree(const Node* node, const std::map<std::string, double>& variables) {
    if (node == nullptr) {
        throw std::runtime_error("Null node encountered");
    }
    
    // Number
    if (node->isNumber()) {
        return std::get<double>(node->value);
    }
    
    // Variable
    if (node->isVariable()) {
        std::string varName = std::get<std::string>(node->value);
        auto it = variables.find(varName);
        if (it == variables.end()) {
            throw std::runtime_error("Undefined variable: " + varName);
        }
        return it->second;
    }
    
    // Operator - evaluate children
    std::string op = std::get<std::string>(node->value);
    
    double leftVal = 0.0;
    double rightVal = 0.0;
    
    if (node->left) {
        leftVal = evaluateTree(node->left.get(), variables);
    }
    
    if (node->right) {
        rightVal = evaluateTree(node->right.get(), variables);
    }
    
    // Apply operator
    if (op == "+") {
        return leftVal + rightVal;
    } else if (op == "-") {
        return leftVal - rightVal;
    } else if (op == "*") {
        return leftVal * rightVal;
    } else if (op == "/") {
        if (rightVal == 0.0) {
            throw std::runtime_error("Division by zero");
        }
        return leftVal / rightVal;
    } else if (op == "^") {
        return std::pow(leftVal, rightVal);
    } else if (op == "√") {
        if (leftVal < 0.0) {
            throw std::runtime_error("Square root of negative number");
        }
        return std::sqrt(leftVal);
    } else {
        throw std::runtime_error("Unknown operator: " + op);
    }
}